src/net/FetchHttpClient.ts

```diff
@@ -15,19 +21,95 @@ import {
-  _fetchFn?: typeof fetch;
+  private readonly _fetchFn: FetchWithTimeout;
 
   constructor(fetchFn?: typeof fetch) {
     super();
-    this._fetchFn = fetchFn;
+
+    // Default to global fetch if available
+    if (!fetchFn) {
+      if (!globalThis.fetch) {
+        throw new Error(
+          'fetch() function not provided and is not defined in the global scope. ' +
+            'You must provide a fetch implementation.'
+        );
+      }
+      fetchFn = globalThis.fetch;
+    }
+
+    // Both timeout behaviors differs from Node:
+    // - Fetch uses a single timeout for the entire length of the request.
+    // - Node is more fine-grained and resets the timeout after each stage of the request.
+    if (globalThis.AbortController) {
+      // Utilise native AbortController if available
+      // AbortController was added in Node v15.0.0, v14.17.0
+      this._fetchFn = FetchHttpClient.makeFetchWithAbortTimeout(fetchFn);
+    } else {
+      // Fall back to racing against a timeout promise if not available in the runtime
+      // This does not actually cancel the underlying fetch operation or resources
+      this._fetchFn = FetchHttpClient.makeFetchWithRaceTimeout(fetchFn);
+    }
+  }
+
+  private static makeFetchWithRaceTimeout(
+    fetchFn: typeof fetch
+  ): FetchWithTimeout {
+    return (url, init, timeout): Promise<Response> => {
+      let pendingTimeoutId: NodeJS.Timeout | null;
+      const timeoutPromise = new Promise<never>((_, reject) => {
+        pendingTimeoutId = setTimeout(() => {
+          pendingTimeoutId = null;
+          reject(HttpClient.makeTimeoutError());
+        }, timeout);
+      });
+
+      const fetchPromise = fetchFn(url, init);
+      return Promise.race([fetchPromise, timeoutPromise]).finally(() => {
+        if (pendingTimeoutId) {
+          clearTimeout(pendingTimeoutId);
+        }
+      });
+    };
+  }
+
+  private static makeFetchWithAbortTimeout(
+    fetchFn: typeof fetch
+  ): FetchWithTimeout {
+    return async (url, init, timeout): Promise<Response> => {
+      // Use AbortController because AbortSignal.timeout() was added later in Node v17.3.0, v16.14.0
+      const abort = new AbortController();
+      let timeoutId: NodeJS.Timeout | null = setTimeout(() => {
+        timeoutId = null;
+        abort.abort(HttpClient.makeTimeoutError());
+      }, timeout);
+      try {
+        return await fetchFn(url, {
+          ...init,
+          signal: abort.signal,
+        });
+      } catch (err) {
+        // Some implementations, like node-fetch, do not respect the reason passed to AbortController.abort()
+        // and instead it always throws an AbortError
+        // We catch this case to normalise all timeout errors
+        if ((err as any).name === 'AbortError') {
+          throw HttpClient.makeTimeoutError();
+        } else {
+          throw err;
+        }
+      } finally {
+        if (timeoutId) {
```

Comment:
nit: is this ever not true?

Reply:
Yeah if the timeout occurs before the fetch resolves then it is unnecessary to clear the timeout, and `timeoutId` will have been set to `null` in the timout callback.

Normally I wouldn't bother with setting `timeoutId` to `null` and doing this check, as I'm pretty sure calling `clearTimeout` after it has already fired is just a no-op [^1]. But I was just following suit with how the `Promise.race()` timeout was originally implemented here.

[^1]: The web spec states that [it is guaranteed that a timeoutID value will never be reused](https://developer.mozilla.org/en-US/docs/Web/API/setTimeout#return_value), but I can't find a similar assertion in the Node.js docs, so perhaps it is best to stick with this method.